engine/src/main/java/org/terasology/rendering/dag/RenderTaskListGenerator.java

```diff
@@ -25,29 +28,48 @@
  */
 public final class RenderTaskListGenerator {
     private static final Logger logger = LoggerFactory.getLogger(RenderTaskListGenerator.class);
+    private List<Object> intermediateList;
+    private List<RenderPipelineTask> taskList;
 
     public RenderTaskListGenerator() {
+        taskList = Lists.newArrayList();
+        intermediateList = Lists.newArrayList();
     }
 
     public List<RenderPipelineTask> generateFrom(List<Node> orderedNodes) {
-        logger.info("Task List: "); // TODO: remove in the future or turn it into debug
-        List<Object> intermediateList = generateIntermediateList(orderedNodes);
-        return generateTaskList(intermediateList);
+        generateIntermediateList(orderedNodes);
+        logList("-- Intermediate List --", intermediateList); // TODO: remove in the future or turn it into debug
+        generateTaskList();
+        logList("-- Task List --", taskList); // TODO: remove in the future or turn it into debug
+        return taskList;
     }
 
-    private List<RenderPipelineTask> generateTaskList(List<Object> intermediateList) {
-        List<RenderPipelineTask> taskList = Lists.newArrayList();
-
-        for (Object o : intermediateList) {
-            taskList.add(generateTask(o));
+    private void logList(String title, List<?> list) {
+        logger.info(title);
+        for (Object o : list) {
+            logger.info(o.toString());
         }
-
-        return taskList;
     }
 
-    private List<Object> generateIntermediateList(List<Node> orderedNodes) {
-        List<Object> intermediateList = Lists.newArrayList();
+    private void generateTaskList() {
+        taskList.clear();
+
+        Map map = Maps.newHashMap();
+        for (Object object : intermediateList) {
+            if (object instanceof StateChange) {
+                map.put(object.getClass(), object);
+            } else {
+                for (Object o : map.values()) {
+                    taskList.add(((StateChange) o).generateTask());
+                }
+                map.clear();
+                taskList.add(new NodeTask((Node) object));
+            }
+        }
+    }
```

Comment:
Hmmmm... I almost wrote this method off but I think it has its merits.

HOWEVER. =)

1. After the last node the list effectively loops back to the beginning (in the next frame), and there some defaults might be needed. So, after the last node we should make sure to have the necessary tasks to set values back to defaults UNLESS they'd be redundant given the state changes required before the first NodeTask gets executed.

2. This method is capable of eliminating redundant tasks _between_ NodeTasks.  However, take the following intermediate list:

- bind fbo "bar" 
- process node ALPHA
- bind fbo "display"
- bind fbo "bar"
- process node BRAVO
- bind fbo "display"
- bind fbo "bar"
- process node CHARLIE
- bind fbo "display"

The two redundant "bind fbo display" state changes are correctly handled and their tasks do not make it into the tasklist. But the second and third "bind fbo bar" are also redundant and their tasks would unnecessarily go into the tasklist.

So, I think you need to store StateChanges that are -not- redundant in another map. When you iterate over the state changes that are found to be -not- redundant _between nodes_, you need to check if they are redundant with the state change of the same subtype stored in the map. If it isn't redundant the state change can get its task into the tasklist AND it replaces the state change of the same subtype in the map. If it is redundant it should be ignored.

And if you want to add the cherry on top (not strictly necessary I think), a NON redundant back-to-default state change should get into the tasklist but should NOT get stored into the "outer" map. Rather, it should cause the state change of the same subtype in the outer map, if any, to be removed. This would keep the map smaller.

If it isn't clear the "outer map" is the one this method implementation doesn't have yet, while the "inner map" is the one you are already using to generate state change tasks between nodes. Perhaps you could call them "persistingStateChanges" (outer) and "intranodeStateChanges".